File: app/application/repositories/snapshot_analysis.py
"""Repository for SnapshotAnalysis persistence.

Provides CRUD operations and review-queue queries for the ML analysis
results stored in the ``snapshot_analyses`` table.
"""

from datetime import datetime

from sqlalchemy import delete, func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.models import SnapshotAnalysis


class SnapshotAnalysisRepository:
    """Provide CRUD and query access to SnapshotAnalysis records.

    Args:
        session: The async SQLAlchemy session to use for all operations.
    """

    def __init__(self, session: AsyncSession):
        self._session = session

    async def delete_by_snapshot_ids(self, snapshot_ids: list[int]) -> int:
        """Delete all analysis records for the given snapshot IDs.

        Args:
            snapshot_ids: Snapshot primary keys whose analyses should be removed.

        Returns:
            The number of analysis records deleted.
        """
        if not snapshot_ids:
            return 0
        result = await self._session.execute(
            delete(SnapshotAnalysis).where(SnapshotAnalysis.snapshot_id.in_(snapshot_ids))
        )
        await self._session.flush()
        return result.rowcount or 0

    async def add(self, analysis: SnapshotAnalysis) -> None:
        """Persist a new analysis result to the database.

        Args:
            analysis: The unsaved SnapshotAnalysis instance.
        """
        self._session.add(analysis)

    async def get_by_id(self, analysis_id: int) -> SnapshotAnalysis | None:
        """Fetch a single analysis record by its primary key.

        Args:
            analysis_id: The unique identifier of the analysis.

        Returns:
            The SnapshotAnalysis if found, otherwise None.
        """
        result = await self._session.execute(
            select(SnapshotAnalysis).where(SnapshotAnalysis.id == analysis_id)
        )
        return result.scalar_one_or_none()

    async def get_by_snapshot(self, snapshot_id: int, model_name: str | None = None) -> list[SnapshotAnalysis]:
        """Find all analyses for a given snapshot, optionally filtered by model.

        Args:
            snapshot_id: The snapshot to query analyses for.
            model_name: Optional model name filter (e.g. ``yolov8n``).

        Returns:
            List of matching SnapshotAnalysis records (may be empty).
        """
        stmt = select(SnapshotAnalysis).where(SnapshotAnalysis.snapshot_id == snapshot_id)
        if model_name:
            stmt = stmt.where(SnapshotAnalysis.model_name == model_name)
        result = await self._session.execute(stmt)
        return list(result.scalars().all())

    async def get_pending_reviews(self, limit: int = 5000) -> list[SnapshotAnalysis]:
        """Fetch analyses that are flagged for human review.

        Args:
            limit: Maximum number of items to return.

        Returns:
            Ordered list of SnapshotAnalysis records requiring review.
        """
        result = await self._session.execute(
            select(SnapshotAnalysis)
            .where(SnapshotAnalysis.review_required == True)

            .order_by(SnapshotAnalysis.analyzed_at.desc())
            .limit(limit)
        )
        return list(result.scalars().all())

    async def get_by_camera_and_date(
        self, camera_id: int, snapshot_ids: list[int]
    ) -> list[SnapshotAnalysis]:
        """Fetch analyses for a set of snapshot IDs belonging to one camera.

        Used by the report endpoint to enrich snapshots with analysis data.

        Args:
            camera_id: The camera these snapshots belong to (used for filtering).
            snapshot_ids: List of snapshot primary keys to query.

        Returns:
            List of matching SnapshotAnalysis records ordered by most recent.
        """
        if not snapshot_ids:
            return []
        result = await self._session.execute(
            select(SnapshotAnalysis)
            .where(SnapshotAnalysis.snapshot_id.in_(snapshot_ids))
            .order_by(SnapshotAnalysis.analyzed_at.desc())
        )
        return list(result.scalars().all())

    async def count_pending_reviews(self) -> int:
        """Return the number of analyses flagged for human review.

        Returns:
            Total count of flagged but un-reviewed analyses.
        """
        result = await self._session.execute(
            select(func.count(SnapshotAnalysis.id)).where(
                SnapshotAnalysis.review_required == True,

            )
        )
        return result.scalar() or 0

    async def update_review(
        self, analysis_id: int, review_required: bool, review_reason: str | None = None
    ) -> None:
        """Update the review flag and optional reason for an analysis.

        Args:
            analysis_id: The analysis to update.
            review_required: Whether the snapshot still requires review.
            review_reason: Optional explanation for the review decision.
        """
        values: dict = {
            "review_required": review_required,
            "updated_at": datetime.now(),
        }
        if review_reason is not None:
            values["review_reason"] = review_reason
        await self._session.execute(
            update(SnapshotAnalysis).where(SnapshotAnalysis.id == analysis_id).values(**values)
        )

    async def get_detections(
        self,
        days_back: int = 1,
        camera_id: int | None = None,
        class_name: str | None = None,
        limit: int = 500,
        offset: int = 0,
        date_from: str | None = None,
    ) -> list[tuple]:
        """Fetch all analyses with detections, joined with snapshot and camera data.

        Args:
            days_back: Only include analyses from the last N days.
            camera_id: Optional camera ID filter.
            class_name: Optional object class filter (JSON text search).
            limit: Maximum rows to return.
            offset: Pagination offset.
            date_from: Specific date (YYYY-MM-DD) — overrides days_back.

        Returns:
            List of tuples with (analysis fields, camera name, snapshot fields).
        """
        from datetime import datetime, timedelta

        if date_from:
            since = datetime.strptime(date_from, "%Y-%m-%d")
            until = since + timedelta(days=1)
        else:
            now = datetime.now()
            if days_back == 0:
                since = now.replace(hour=0, minute=0, second=0, microsecond=0)
            else:
                since = now - timedelta(days=days_back)
            until = None
        from app.domain.models import Snapshot, Camera

        stmt = (
            select(
                SnapshotAnalysis,
                Camera.name,
                Snapshot.captured_at,
                Snapshot.image_path,
                Snapshot.camera_id,
            )
            .join(Snapshot, SnapshotAnalysis.snapshot_id == Snapshot.id)
            .join(Camera, Snapshot.camera_id == Camera.id)
            .where(
                SnapshotAnalysis.objects_json.isnot(None),
                SnapshotAnalysis.analyzed_at >= since,
            )
            .order_by(SnapshotAnalysis.analyzed_at.desc())
            .offset(offset)
            .limit(limit)
        )
        if until is not None:
            stmt = stmt.where(SnapshotAnalysis.analyzed_at < until)
        if camera_id is not None:
            stmt = stmt.where(Snapshot.camera_id == camera_id)
        if class_name is not None:
            stmt = stmt.where(SnapshotAnalysis.objects_json.contains(class_name))

        result = await self._session.execute(stmt)
        return list(result.all())

File: internal/repository/snapshot_analysis.go
package repository

import (
	"context"
	"time"

	"github.com/jmoiron/sqlx"

	"github.com/Lmex89/home-cameras/internal/domain"
)

// SnapshotAnalysisRepository persists ML analysis results.
type SnapshotAnalysisRepository struct {
	db DBTX
}

// NewSnapshotAnalysisRepository builds a snapshot analysis repository.
func NewSnapshotAnalysisRepository(db DBTX) *SnapshotAnalysisRepository {
	return &SnapshotAnalysisRepository{db: db}
}

const analysisCols = `id, snapshot_id, model_name, model_version, status, objects_json,
	person_count, review_required, review_reason, anomaly_score, error_message,
	analyzed_at, created_at, updated_at`

// Add inserts a new analysis row and populates its id.
func (r *SnapshotAnalysisRepository) Add(ctx context.Context, a *domain.SnapshotAnalysis) error {
	res, err := r.db.ExecContext(ctx,
		`INSERT INTO snapshot_analyses (snapshot_id, model_name, model_version, status,
			objects_json, person_count, review_required, review_reason, anomaly_score,
			error_message, analyzed_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		a.SnapshotID, a.ModelName, a.ModelVersion, a.Status, a.ObjectsJSON,
		a.PersonCount, a.ReviewRequired, a.ReviewReason, a.AnomalyScore,
		a.ErrorMessage, a.AnalyzedAt)
	if err != nil {
		return err
	}
	id, err := res.LastInsertId()
	if err != nil {
		return err
	}
	a.ID = id
	return nil
}

// GetByID returns one analysis.
func (r *SnapshotAnalysisRepository) GetByID(ctx context.Context, id int64) (*domain.SnapshotAnalysis, error) {
	var a domain.SnapshotAnalysis
	err := sqlx.GetContext(ctx, r.db, &a,
		"SELECT "+analysisCols+" FROM snapshot_analyses WHERE id = ?", id)
	if err != nil {
		return nil, err
	}
	return &a, nil
}

// GetBySnapshot returns analyses for a snapshot, optionally by model.
func (r *SnapshotAnalysisRepository) GetBySnapshot(ctx context.Context, snapshotID int64, modelName ...string) ([]domain.SnapshotAnalysis, error) {
	query := "SELECT " + analysisCols + " FROM snapshot_analyses WHERE snapshot_id = ?"
	args := []any{snapshotID}
	if len(modelName) > 0 && modelName[0] != "" {
		query += " AND model_name = ?"
		args = append(args, modelName[0])
	}
	var out []domain.SnapshotAnalysis
	err := sqlx.SelectContext(ctx, r.db, &out, query, args...)
	return out, err
}

// GetPendingReviews returns analyses flagged for human review.
func (r *SnapshotAnalysisRepository) GetPendingReviews(ctx context.Context, limit int) ([]domain.SnapshotAnalysis, error) {
	var out []domain.SnapshotAnalysis
	err := sqlx.SelectContext(ctx, r.db, &out,
		"SELECT "+analysisCols+` FROM snapshot_analyses
		 WHERE review_required = 1
		 ORDER BY analyzed_at DESC LIMIT ?`, limit)
	return out, err
}

// GetByCameraAndDate returns analyses for a set of snapshot IDs.
func (r *SnapshotAnalysisRepository) GetByCameraAndDate(ctx context.Context, snapshotIDs []int64) ([]domain.SnapshotAnalysis, error) {
	if len(snapshotIDs) == 0 {
		return nil, nil
	}
	query, args, err := sqlx.In(
		"SELECT "+analysisCols+` FROM snapshot_analyses
		 WHERE snapshot_id IN (?) ORDER BY analyzed_at DESC`, snapshotIDs)
	if err != nil {
		return nil, err
	}
	var out []domain.SnapshotAnalysis
	err = sqlx.SelectContext(ctx, r.db, &out, r.db.Rebind(query), args...)
	return out, err
}

// CountPendingReviews counts flagged analyses.
func (r *SnapshotAnalysisRepository) CountPendingReviews(ctx context.Context) (int, error) {
	var n int
	err := sqlx.GetContext(ctx, r.db, &n,
		"SELECT COUNT(id) FROM snapshot_analyses WHERE review_required = 1")
	return n, err
}

// UpdateReview sets the review flag and optional reason.
func (r *SnapshotAnalysisRepository) UpdateReview(ctx context.Context, id int64, required bool, reason *string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE snapshot_analyses SET review_required = ?, review_reason = ?,
		 updated_at = CURRENT_TIMESTAMP WHERE id = ?`, required, reason, id)
	return err
}

// GetDetections returns analyses with detections joined to snapshots
// and cameras, mirroring the legacy detections browser query.
func (r *SnapshotAnalysisRepository) GetDetections(ctx context.Context, since, until *time.Time, cameraID int64, className string, limit, offset int) ([]DetectionRow, error) {
	query := `SELECT sa.id AS analysis_id, sa.snapshot_id, s.camera_id,
		c.name AS camera_name, s.captured_at, s.image_path,
		sa.model_name, sa.objects_json, sa.person_count,
		sa.review_required, sa.review_reason, sa.analyzed_at
		FROM snapshot_analyses sa
		JOIN snapshots s ON s.id = sa.snapshot_id
		JOIN cameras c ON c.id = s.camera_id
		WHERE sa.objects_json IS NOT NULL`
	args := make([]any, 0, 6)
	if since != nil {
		query += " AND sa.analyzed_at >= ?"
		args = append(args, since.Format("2006-01-02 15:04:05"))
	}
	if until != nil {
		query += " AND sa.analyzed_at < ?"
		args = append(args, until.Format("2006-01-02 15:04:05"))
	}
	if cameraID > 0 {
		query += " AND s.camera_id = ?"
		args = append(args, cameraID)
	}
	if className != "" {
		query += " AND sa.objects_json LIKE ?"
		args = append(args, "%"+className+"%")
	}
	query += " ORDER BY sa.analyzed_at DESC LIMIT ? OFFSET ?"
	args = append(args, limit, offset)

	var rows []DetectionRow
	if err := sqlx.SelectContext(ctx, r.db, &rows, query, args...); err != nil {
		return nil, err
	}
	return rows, nil
}

// DeleteBySnapshotIDs deletes analyses for the given snapshots.
func (r *SnapshotAnalysisRepository) DeleteBySnapshotIDs(ctx context.Context, ids []int64) (int64, error) {
	if len(ids) == 0 {
		return 0, nil
	}
	query, args, err := sqlx.In("DELETE FROM snapshot_analyses WHERE snapshot_id IN (?)", ids)
	if err != nil {
		return 0, err
	}
	res, err := r.db.ExecContext(ctx, r.db.Rebind(query), args...)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

// DetectionRow is a flattened detection result for the API.
type DetectionRow struct {
	AnalysisID     int64          `db:"analysis_id" json:"analysis_id"`
	SnapshotID     int64          `db:"snapshot_id" json:"snapshot_id"`
	CameraID       int64          `db:"camera_id" json:"camera_id"`
	CameraName     string         `db:"camera_name" json:"camera_name"`
	CapturedAt     domain.SQLTime `db:"captured_at" json:"captured_at"`
	ImagePath      string         `db:"image_path" json:"image_path"`
	ModelName      string         `db:"model_name" json:"model_name"`
	ObjectsJSON    *string        `db:"objects_json" json:"objects_json"`
	PersonCount    int            `db:"person_count" json:"person_count"`
	ReviewRequired bool           `db:"review_required" json:"review_required"`
	ReviewReason   *string        `db:"review_reason" json:"review_reason"`
	AnalyzedAt     domain.SQLTime `db:"analyzed_at" json:"analyzed_at"`
}
